### PR TITLE
[JW8-5170] Ad and Chapters Cues Should Both Appear on Timeline

### PR DESCRIPTION
**Pull request by waxidiotic**
JW8-5170

### This PR will...

* Create new `getCues()` API method that returns the list of cues on the model.
* Create new `addCues()` API method that gets the current cues and appends new cues.
* Make sure chapter cues are set to the model.
* Also call `resetChapters()` inside of `reset()` so it occurs on each playlist item

### Why is this Pull Request needed?

* When ad plugins were adding cues at different times than chapter cues, whatever cues were added first were cleared.

### Are there any points in the code the reviewer needs to double check?

n/a

### Are there any Pull Requests open in other repos which need to be merged with this?

https://github.com/jwplayer/jwplayer-ads-vast/pull/485
https://github.com/jwplayer/jwplayer-ads-googima/pull/466
https://github.com/jwplayer/jwplayer-ads-freewheel/pull/152

#### Addresses Issue(s):

JW8-5170

### CODE REVIEW

**Comment by waxidiotic** [comment]
The one failure can't be reproduced in PR build's harness.

**Comment by johnBartos** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  🛠 There are modified src files, but no test changes. Add tests if you're able to.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by aizaztoppa** [comment]
test this please

File: src/js/view/controls/components/timeslider.js [hunk 1]
```diff
@@ -280,7 +280,6 @@ class TimeSlider extends Slider {
     }
 
     addCues(model, cues) {
-        this.resetChapters();
```
Comment (pajong) [inline src/js/view/controls/components/timeslider.js]: I don't think this is safe. 
`api.setCues` changes the `cues` in `model`, which then calls this function. This function used to clear the already added cues. Removing this code will change the api behavior.

Comment (waxidiotic) [reply src/js/view/controls/components/timeslider.js]: @robwalch - Your thoughts on this? @pajong and I were talking. He's worried that is a customer calls `setCues()` which isn't documented, it would cause duplicate cues. But, we have to make sure that multiple parts of the codebase are able to call `setCues()` (i.e. ads and chapters).

Comment (waxidiotic) [reply src/js/view/controls/components/timeslider.js]: New thought: could change `setCues()` in controller.js to clear the cues if there are any before setting them. I believe the only thing that would be able to call `setCues()` would be the JavaScript API. The mixins and ad plugins themselves don't call it.

Comment (pajong) [reply src/js/view/controls/components/timeslider.js]: The ads plugins call `setCues()`
e.g.
https://github.com/jwplayer/jwplayer-ads-vast/blob/master/src/js/vastPlugin.js#L468
https://github.com/jwplayer/jwplayer-ads-googima/blob/master/src/PlaylistItemManager.js#L181

Comment (waxidiotic) [reply src/js/view/controls/components/timeslider.js]: Yeah, I already realized that it wouldn't work. :(

Comment (waxidiotic) [reply src/js/view/controls/components/timeslider.js]: @jnatalzia - Curious your thoughts on this too.

Comment (jnatalzia) [reply src/js/view/controls/components/timeslider.js]: I don't have too much context on exactly what `resetChapters` here means but what's the risk around leaving it here to preserve existing functionality and adding it to `reset`?

Comment (waxidiotic) [reply src/js/view/controls/components/timeslider.js]: @jnatalzia - It would cause the issue that is the cause of this work which is that the chapter cues will be cleared when the ad plugins call `addCues()`.

Comment (robwalch) [reply src/js/view/controls/components/timeslider.js]: @waxidiotic I am concerned that this only works if setCues is only called twice when you have chapter cues and adds. Now you can't use `setCues` to reset the cues (API change which Jong pointed out.)

What we should do is add a `getCues` method if we don't have one, and the ads or chapters should get cues, add their cues to a copy of that array, and then set cues to the new updated list.

**Comment by waxidiotic** [comment]
test this please

**Comment by waxidiotic** [comment]
test this please

File: src/js/view/controls/components/chapters.mixin.js [hunk 2]
```diff
@@ -10,6 +10,10 @@ class Cue {
     }
 
     align(duration) {
+        if (!this.time) {
+            return;
+        }
```
Comment (robwalch) [inline src/js/view/controls/components/chapters.mixin.js]: Why is this necessary?

`this.time = time` is set in the constructor. You should never be instantiating cues without a time. Isn't `0` still valid?

File: src/js/view/controls/components/chapters.mixin.js [hunk 3]
```diff
@@ -35,15 +39,29 @@ const ChaptersMixin = {
         if (Array.isArray(data)) {
             data.forEach((obj) => this.addCue(obj));
             this.drawCues();
+            const existingCues = this._model.get('cues');
+            const playerCues = existingCues.concat(data);
+            this._model.set('cues', playerCues);
```
Comment (robwalch) [inline src/js/view/controls/components/chapters.mixin.js]: Why do you draw the cues and then update the model?

File: src/js/view/controls/components/chapters.mixin.js [hunk 4]
```diff
@@ -35,15 +39,29 @@ const ChaptersMixin = {
...
     },
 
     chaptersFailed: function () {},
 
     addCue: function (obj) {
+        this.cues.forEach((cue) => {
+            if (cue.time === obj.begin) {
+                this.removeCue(cue);
+            }
+        });
```
Comment (robwalch) [inline src/js/view/controls/components/chapters.mixin.js]: We didn't need to bother with duplicates before. Why do we now, and why is the chapter mixin responsible for this? Why are chapter cues more important than other cues?

File: src/js/view/controls/components/chapters.mixin.js [hunk 5]
```diff
@@ -35,15 +39,29 @@ const ChaptersMixin = {
...
         this.cues.push(new Cue(obj.begin, obj.text));
     },
 
+    removeCue: function (cueToRemove) {
```
Comment (robwalch) [inline src/js/view/controls/components/chapters.mixin.js]: This doesn't need to be its own function exposed on the object. Update the model, then reset and redraw all cues.

Comment (robwalch) [reply src/js/view/controls/components/chapters.mixin.js]: > Fix a JS exception that was occurring when clearing out the queues using `jwplayer().setCues([{}]);`

Can't you just call `jwplayer().setCues([]);`? `{}` is not a valid cue. What calls it? Don't add a check to each cue, when you can handle this at a higher level and just reset.

> Prevent duplicate cues at the same time on the timeslider
Cues with the same time but different content are not duplicates. Establish rules about which types of cues take priority. Handle this when drawing the full list of cues not by removing and replacing elements one at a time.

File: src/js/view/controls/components/chapters.mixin.js (hunk 3 above)
Comment (waxidiotic) [reply src/js/view/controls/components/chapters.mixin.js]: I'm not sure why it would matter. Can you explain?

File: src/js/view/controls/components/chapters.mixin.js (hunk 4 above)
Comment (waxidiotic) [reply src/js/view/controls/components/chapters.mixin.js]: You told me last week that we don't want duplicate cues to appear if there are two cues set at the same time. Also, it's not necessarily chapters over ads - it's just what gets added first.

File: src/js/view/controls/components/chapters.mixin.js (hunk 5 above)
Comment (waxidiotic) [reply src/js/view/controls/components/chapters.mixin.js]: @robwalch 
The question I have would be how to tell what is adding cues. It just happens to be that chapters are added when `playlistItem` fires while the ad plugins add them when the playback is initiated. I assume that we would always want ad cues to take priority over chapter cues, right? If so, I can check in `drawCues()`.

This is a little stream of consciousness but I want to make sure I'm thinking along the same lines.

File: src/js/view/controls/components/chapters.mixin.js (hunk 4 above)
Comment (waxidiotic) [reply src/js/view/controls/components/chapters.mixin.js]: I removed this. Will handle dupes in `drawCues()`.

File: src/js/view/controls/components/chapters.mixin.js (hunk 3 above)
Comment (waxidiotic) [reply src/js/view/controls/components/chapters.mixin.js]: Refactored this to just set the cue to the model when the chapters are loaded. Changing the cues on the model will then call addCue() and drawCues().

File: src/js/view/controls/components/chapters.mixin.js (hunk 2 above)
Comment (waxidiotic) [reply src/js/view/controls/components/chapters.mixin.js]: This was because I was calling `setCues()` wrong to test.

File: src/js/controller/controller.js [hunk 6]
```diff
@@ -978,9 +978,19 @@ Object.assign(Controller.prototype, {
...
         this.setCues = function (cues) {
-            _model.set('cues', cues);
+            let newCues = [];
+            if (cues && cues.length) {
+                // If setCues() is called with empty array, we want to clear the cues.
+                // Otherwise, add the new cues to the existing cues.
+                // This allows setting of cues by ad plugins or setCues() API.
+                const existingCues = _model.get('cues');
+                newCues = existingCues ? existingCues.concat(cues) : cues;
+            }
```
Comment (robwalch) [inline src/js/controller/controller.js]: This goes back to what the original feedback @pajong and I left - `setCues` should not append to cues.

You could add a new method called `addCues` that does this, and then avoid the repetition of calling `setCues(getCues().append(newCues))` in every place that currently calls `setCues()` to add chapter or ads markers.

File: src/js/view/controls/components/chapters.mixin.js [hunk 7]
```diff
@@ -33,8 +33,8 @@ const ChaptersMixin = {
     chaptersLoaded: function (evt) {
         const data = srt(evt.responseText);
         if (Array.isArray(data)) {
-            data.forEach((obj) => this.addCue(obj));
-            this.drawCues();
+            // Add chapter cues directly to model which will trigger addCue()
+            this._model.set('cues', data);
```
Comment (robwalch) [inline src/js/view/controls/components/chapters.mixin.js]: This should append the new cues, or call a new `addCues` method which does the appending. Since you have direct access to the model, you could slice model cues, append and set, and then depend on the `change:cues` event to handle the update in the time slider chapters mixin.

**Comment by waxidiotic** [comment]
test this please

File: src/js/controller/controller.js [hunk 8]
```diff
@@ -977,10 +977,18 @@ Object.assign(Controller.prototype, {
             _programController.controls = mode;
         };
 
+        this.addCues = function (cues) {
+            const existingCues = this.getCues();
+            const newCues = existingCues ? existingCues.concat(cues) : cues;
```
Comment (robwalch) [inline src/js/controller/controller.js]: 1. Let's default the cues in model to an empty array so that we don't need this ternary operator.
2. In `setCues` let's validate the input - it must be an array, otherwise we should not update the model.

File: src/js/view/controls/components/chapters.mixin.js [hunk 9]
```diff
@@ -33,8 +33,10 @@ const ChaptersMixin = {
     chaptersLoaded: function (evt) {
         const data = srt(evt.responseText);
         if (Array.isArray(data)) {
-            data.forEach((obj) => this.addCue(obj));
-            this.drawCues();
+            // Add chapter cues directly to model which will trigger addCue()
+            const existingCues = this._model.get('cues');
+            const newCues = existingCues ? existingCues.concat(data) : data;
```
Comment (robwalch) [inline src/js/view/controls/components/chapters.mixin.js]: Guarantee that model 'cues' is always an array so you don't have to check if existingCues is defined (see above)

File: src/js/view/controls/components/timeslider.js [hunk 10]
```diff
@@ -280,12 +280,14 @@ class TimeSlider extends Slider {
...
+        // Only reset chapters if setCues() is called with empty array (to clear cues)
```
Comment (robwalch) [inline src/js/view/controls/components/timeslider.js]: 1. Can we rename `addCues` to set or update cues?
2. Shouldn't you always reset (`resetChapters` is really `resetCues` no?) before adding each new cue?

**Comment by waxidiotic** [comment]
Replaced with https://github.com/jwplayer/jwplayer/pull/3313
